pkg/sentry/fs/fsutil/cache_reclaim.go

```diff
@@ -0,0 +1,350 @@
+package fsutil
+
+import (
+        "sync"
+        "sync/atomic"
+        "time"
+        "runtime"
+
+        "gvisor.dev/gvisor/pkg/ilist"
+        "gvisor.dev/gvisor/pkg/log"
+        "gvisor.dev/gvisor/pkg/sentry/context"
+        "gvisor.dev/gvisor/pkg/sentry/memmap"
+        "gvisor.dev/gvisor/pkg/sentry/pgalloc"
+        "gvisor.dev/gvisor/pkg/sentry/platform"
+        "gvisor.dev/gvisor/pkg/sentry/usermem"
+)
+
+const (
+        LRU_ACTIVE_LIST = 0
+        LRU_INACTIVE_LIST = 1
+        LRU_LIST_NUM = 2
+)
+
+type LruEntry struct {
+	ilist.Entry
+	fr platform.FileRange
+	mr memmap.MappableRange
+	c *CachingInodeOperations
+	accessed bool
+	lru int
+}
+
+type LruManager struct {
+	destroyed bool
+	forceReclaim bool
+
+	// LRULists store the current file ranges consumed by all the files
+        // with the minimum granularity of 4K
+        // LRUMap records the mappings from the start of file range to its entry
+        //
+        // LRULists and LRUMap are protected by LRUMapMu
+        lists [LRU_LIST_NUM]ilist.List
+	length [LRU_LIST_NUM]int
+        mappings map[uint64]*LruEntry
+        lruMu sync.RWMutex
+
+	f *pgalloc.MemoryFile
+}
+
+func NewLruEntry(fr platform.FileRange, mr memmap.MappableRange, c *CachingInodeOperations) *LruEntry {
+        return &LruEntry{fr: fr, mr: mr, c: c, accessed: true, lru: LRU_ACTIVE_LIST}
+}
+
+func update(e *LruEntry, fr platform.FileRange, mr memmap.MappableRange, c *CachingInodeOperations) {
+	e.fr = fr
+	e.mr = mr
+	e.c = c
+	e.accessed = true
+}
+
+var mgrMu sync.Mutex
+var mgrMap map[*pgalloc.MemoryFile]*LruManager = make(map[*pgalloc.MemoryFile]*LruManager)
+
+func NewLruManager(interval time.Duration, f *pgalloc.MemoryFile) *LruManager {
+	// Created only once for each MemoryFile
```

Comment:
See comment above, I think this memoizing trick actually points that the structure isn't quite right.

Reply:
OK, should we put all these things into pgalloc?